qiskit/algorithms/time_evolvers/variational/variational_principles/real_mc_lachlan_principle.py

```diff
@@ -0,0 +1,151 @@
+#
+# (C) Copyright IBM 2021, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Class for a Real McLachlan's Variational Principle."""
+from __future__ import annotations
+
+import numpy as np
+from numpy import real
+
+from qiskit import QuantumCircuit
+from qiskit.algorithms import AlgorithmError
+from qiskit.algorithms.gradients import (
+    BaseEstimatorGradient,
+    BaseQFI,
+    LinCombQFI,
+)
+from qiskit.algorithms.gradients.lin_comb_estimator_gradient import (
+    DerivativeType,
+    LinCombEstimatorGradient,
+)
+from qiskit.circuit import Parameter
+from qiskit.opflow import (
+    PauliSumOp,
+)
+from qiskit.primitives import Estimator
+from qiskit.quantum_info import Pauli, SparsePauliOp
+from qiskit.quantum_info.operators.base_operator import BaseOperator
+from .real_variational_principle import (
+    RealVariationalPrinciple,
+)
+
+
+class RealMcLachlanPrinciple(RealVariationalPrinciple):
+    """Class for a Real McLachlan's Variational Principle. It aims to minimize the distance
+    between both sides of the Schrödinger equation with a quantum state given as a parametrized
+    trial state. The principle leads to a system of linear equations handled by a linear solver.
+    The real variant means that we consider real time dynamics.
+    """
+
+    def __init__(
+        self,
+        qfi: BaseQFI | None = None,
+        gradient: BaseEstimatorGradient | None = None,
+    ) -> None:
+        """
+        Args:
+            qfi: Instance of a class used to compute the QFI. If ``None`` provided, ``LinCombQFI``
+                is used.
+            gradient: Instance of a class used to compute the state gradient. If ``None`` provided,
+                ``LinCombEstimatorGradient`` is used.
+        """
+        # TODO validate that gradient passed has IMAG
+        if gradient is not None and gradient._estimator is not None and qfi is None:
+            estimator = gradient._estimator
+            qfi = LinCombQFI(estimator)
+        elif qfi is None and gradient is None:
+            estimator = Estimator()
+            qfi = LinCombQFI(estimator)
+            gradient = LinCombEstimatorGradient(estimator, derivative_type=DerivativeType.IMAG)
+        self._energy_param = None
+        self._energy = None
+
+        super().__init__(qfi, gradient)
+
+    def evolution_grad(
+        self,
+        hamiltonian: BaseOperator | PauliSumOp,
+        ansatz: QuantumCircuit,
+        param_dict: dict[Parameter, complex],
+        bind_params: list[Parameter],
+        gradient_params: list[Parameter],
+        param_values: list[complex],
```

Comment:
Some comment here:
* Isn't `param_dict` and `param_values` the same information? Below it seems that `param_dict` is not actually used
* `bind_params` seems also like it's not used 
* I think in almost all cases (at least all I know), we compute the gradient wrt all parameters. Could we make `gradient_params` optional and default to all parameters? The gradient primitives should already support that by just passing `None`